### ISSUE TITLE
Enable Link-Time Optimization (LTO)

### ISSUE THREAD

**Issue by zamazan4ik** [iamzubin/holdem#2]
Hi!

I noticed that in the `Cargo.toml` for the Tauri part Link-Time Optimization (LTO) for the project is not enabled. I suggest switching it on since it will reduce the binary size (always a good thing to have) and will likely improve the application's performance a bit.

I suggest enabling LTO only for the Release builds so as not to sacrifice the developers' experience while working on the project since LTO consumes an additional amount of time to finish the compilation routine. If you think that a regular Release build should not be affected by such a change as well, then I suggest adding an additional `dist` or `release-lto` profile where additionally to regular `release` optimizations LTO will also be added. Such a change simplifies life for maintainers and others interested in the project persons who want to build the most performant version of the application. Using ThinLTO should also help to reduce the build-time overhead with LTO. If we enable it on the Cargo profile level, users, who install the application with `cargo install`, will get the LTO-optimized version "automatically". E.g., check `cargo-outdated` Release [profile](https://github.com/kbknapp/cargo-outdated/blob/master/Cargo.toml#L48).

Basically, it can be enabled with the following lines:

```
[profile.release]
lto = true
```

Tauri official docs also [recommends](https://v1.tauri.app/v1/guides/building/app-size/#rust-build-time-optimizations) using LTO (and other optimization options).

Thank you.

**Comment by iamzubin** [comment]
Thanks for the Suggestion and the detailed Github Issue.
I've enabled it and did see quite a difference, not just in file size, but also made the app run faster.

I'm new to rust and spent a lot of time away from compiled languages, didn't know compiler had multiple optimization options.

I don't see LTO being on for all builds causing errors, I personally don't mind keeping it on for all.
I also didn't see it much ( just a quick google search ) but is there any other drawback to using LTO or have you encountered any?


Thanks a lot.

**Comment by zamazan4ik** [comment]
> I don't see LTO being on for all builds causing errors, I personally don't mind keeping it on for all.

Yeah, for this project you should be fine with LTO without issues. LTO often causes build issues in more difficult for a compiler environment like performing cross-language LTO between libraries that are written in different languages (e.g. LTO for a Rust app with a C++ library) due to various reasons like incompatible LTO formats, less tested compiler paths, etc. Your case is definitely a default one for LTO ;)

> I also didn't see it much ( just a quick google search ) but is there any other drawback to using LTO or have you encountered any?

Actually, there are a couple of them but their importance depends on the context. Sometimes there are "interesting" compiler bugs but they are quite rare nowadays in the LTO field. So I would say no - you shouldn't expect issues with LTO except the increased build time and the increased memory usage during the compilation process (it could be important for memory-tight CI envs). In such cases, using ThinLTO also could help to partially mitigate the issue.
